README.md
# Filesystem

![license](https://img.shields.io/badge/license-MIT-brightGreen.svg)
[![build](https://github.com/originphp/filesystem/workflows/CI/badge.svg)](https://github.com/originphp/filesystem/actions)
[![coverage](https://coveralls.io/repos/github/originphp/filesystem/badge.svg?branch=master)](https://coveralls.io/github/originphp/filesystem?branch=master)

Filesystem includes the `File` and `Folder` classes for working with your filesystem.

## Installation

To install this package

```linux
$ composer require originphp/filesystem
```


## File

he file utility wraps some important functions in an easy to use and predictable way.

To use the File utility add the following to the top of your file.

```php
use Origin\Filesystem\File
```

### Info

> As of 2.0 path was renamed to directory and filename was renamed to name

To get information about a file

```php
$info = File::info('/var/www/config/insurance.csv');
```

Which will return an array like this

```
Array
(
    [name] => insurance.csv
    [directory] => /var/www/config
    [timestamp] => 1560067334
    [size] => 1878
    [extension] => csv
    [type] => text/plain
)
```

### Read

To read a file

```php
$contents = File::read('/path/somefile');
```

### Write

To write to a file


```php
File::write('/path/somefile','data goes here');
```

### Append

To append contents to a file

```php
File::append('/path/somefile','and here.');
```

### Delete

To delete a file

```php
File::delete('/path/somefile');
```

### Exists

To check if file exists

```php
$result = File::exists('/path/somefile');
```

### Tmp

When needing to work with temporary files, use tmp, this will create the file, put the contents and return to you the name of the file with path.

```php
$tmpFile = File::tmp('Some temp data');
```

### Copy

To copy a file

```php
File::copy('/path/somefile','somefile-backup');
File::copy('/path/somefile','/another_path/somefile');
```

### Rename

To rename a file

```php
File::rename('/path/somefile','new_name');
```

### Move

To move a file

```php
File::move('/path/somefile','/another_path/somefile');
```

### Permissions

#### Get Permissions

To get the permissions of a file

```php
$permissions = File::perms('/path/somefile'); // returns 0744
```

#### Changing Permissions (chmod)

To change the permissions of a file.

```php
File::chmod('/path/somefile','www-data');
```

#### Getting the owner of the file

```php
$owner = File::owner('/path/somefile'); // returns root
```

#### Changing Ownership (chown)

To change the ownership of a file

```php
File::chown('/path/somefile','www-data');
```

#### Getting the group

To get the group that the file belongs to.

```php
$group = File::group('/path/somefile'); // returns root
```

#### Changing Group (chgrp)

To change the group that the file belongs to.

```php
File::chgrp('/path/somefile','www-data');
```

## Folder

The folder utility helps you work with folders on your file system.

To use the Folder utility add the following to the top of your file.

```php
use Origin\Filesystem\Folder
```

### Create

To create a folder

```php
Folder::create('/var/www/new_folder');
```

To create a folder recursively

```php
Folder::create('/var/www/level1/level2/level3/new_folder',['recursive'=>true]);
```

To set the permissions on the newly created folder

```php
Folder::create('/var/www/new_folder',['mode'=>0755]);
```

### Delete

To delete a folder

```php
Folder::delete('/var/www/bye-bye')
```

To delete a folder recursively, including all files and sub directories.

```php
Folder::delete('/var/www/docs',['recursive'=>true])
```

### Exists

To check if a directory exists

```php
$result = Folder::exists('/path/somedirectory');
```

### List

> As of version 2.0 released 26.09.20 list results are different, path now includes the filename.

To list all contents of a directory

```php
$results = Folder::list('/path/somedirectory');
```

This will return an array of arrays of `FileObjects`

```php
[
    Origin\Filesystem\FileObject Object
    (
        'name' => 'foo.txt',
        'directory' => '/var/www/my_directory',
        'path' =>  '/var/www/my_directory/foo.txt',
        'extension' => 'txt',
        'timestamp' => 14324234,
        'size' => 1234,
        'type' => 'file'
    )
]
```

When the `FileObject` is converted to a string it will become a path e.g. `/var/www/my_directory/foo.txt`.

```php
foreach(Folder::list('/data') as $file){
    $contents = file_get_contents($file->path); // or (string) $file
}
```

You can also get the listing recursively

```php
$results = Folder::list('/path/somedirectory',['recursive'=>true]);
```

To include directories in the results

```php
$results = Folder::list('/path/somedirectory',['directories'=>true]);
```

### Copy

To copy a directory

```php
Folder::copy('/path/somedir','somedir-backup');
Folder::copy('/path/somedir','/another_path/somedir');
```

### Rename

To rename a directory

```php
Folder::rename('/path/somedir','new_name');
```

### Move

To move a directory

```php
Folder::move('/path/somedir','/another_path/somedir');
```

### Permissions

#### Get Permissions

To get the permissions of a directory.

```php
$permissions = Folder::perms('/path/somedir'); // returns 0744
```

#### Changing Permissions (chmod)

To change the permissions of a directory

```php
Folder::chmod('/path/somedir','www-data');
Folder::chmod('/path/somedir','www-data',['recursive'=>true]); // recursive
```

#### Getting the owner of a directory

```php
$owner = Folder::owner('/path/somedir'); // returns root
```

#### Changing Ownership (chown)

To change the ownership of a directory

```php
Folder::chown('/path/somedir','www-data');
Folder::chown('/path/somedir','www-data',['recursive'=>true]);
```

#### Getting the group

To get the group that a directory belongs to.

```php
$group = Folder::group('/path/somedir'); // returns root
```

#### Changing Group (chgrp)

To change the group that the folder belongs to.

```php
Folder::chgrp('/path/somedir','www-data');
Folder::chgrp('/path/somedir','www-data',['recursive'=>true]);
```


CHANGELOG.md
# Changelog

All notable changes to this project will be documented in this file.

The format is based on [Keep a Changelog](https://keepachangelog.com/en/1.0.0/),
and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0.html).

## [3.1.0] - 2021-04-13

### Changed

- Adjusted to use Defer 3.x

## [3.0.2] - 2021-01-07

## [3.0.1] - 2021-01-06

### Fixed

- Fixed error when copying on PHP 8 if folder already exists

## [3.0.0] - 2021-01-04

### Changed

- Changed minimum PHP version to 7.3
- Changed minimum PHPUnit to 9.2
- Changed defer to use v2

## [2.0.0] - 2020-09-26

### Changed

- Changed Folder::list now returns an array of FileObjects
- Changed Folder::list FileObject uses directory instead of path (BC)
- Changed File::info filename has been renamed to name (BC)
- Changed File::info path has been renamed to directory (BC)

## [1.0.0] - 2019-10-11

This component has been decoupled from the [OriginPHP framework](https://www.originphp.com/).


tests/FileTest.php
<?php
/**
 * OriginPHP Framework
 * Copyright 2018 - 2021 Jamiel Sharief.
 *
 * Licensed under The MIT License
 * The above copyright notice and this permission notice shall be included in all copies or substantial
 * portions of the Software.
 *
 * @copyright   Copyright (c) Jamiel Sharief
 * @link        https://www.originphp.com
 * @license     https://opensource.org/licenses/mit-license.php MIT License
 */

namespace Origin\Test\Utility;

use Origin\Filesystem\File;
use Origin\Filesystem\Exception\NotFoundException;

class FileTest extends \PHPUnit\Framework\TestCase
{
    public function testRead()
    {
        $tmp = sys_get_temp_dir() . '/' . uniqid();
        file_put_contents($tmp, 'Hello World!');
        $this->assertEquals('Hello World!', File::read($tmp));
    }

    public function testReadException()
    {
        $this->expectException(NotFoundException::class);
        File::read('/foo/bar.txt');
    }
 
    public function testInfoException()
    {
        $this->expectException(NotFoundException::class);
        File::info('/foo/bar.txt');
    }

    public function testWrite()
    {
        $tmp = sys_get_temp_dir() . '/' . uniqid();
        $this->assertTrue(File::write($tmp, 'foo'));
        $this->assertEquals('foo', file_get_Contents($tmp));
        $this->assertFalse(File::write(sys_get_temp_dir() . '/does-not-exist/file.txt', 'foo'));
    }
 
    public function testAppend()
    {
        $tmp = sys_get_temp_dir() . '/'  . uniqid();
        $this->assertTrue(File::append($tmp, 'foo'));
        $this->assertEquals('foo', file_get_Contents($tmp));
        $this->assertTrue(File::append($tmp, 'bar'));
        $this->assertEquals('foobar', file_get_Contents($tmp));
        $this->assertFalse(File::append(sys_get_temp_dir() . '/does-not-exist/file.txt', 'foo'));
    }

    public function testExists()
    {
        $tmp = sys_get_temp_dir() . '/' . uniqid();
        File::write($tmp, 'foo');
        $this->assertTrue(File::exists($tmp));
        $this->assertFalse(File::exists('/foo/bar.txt'));
    }

    /**
     * @depends testExists
     *
     * @return void
     */
    public function testDelete()
    {
        $tmp = sys_get_temp_dir() . '/' . uniqid();
        file_put_contents($tmp, 'foo');
        $this->assertTrue(File::delete($tmp));
        $this->expectException(NotFoundException::class);
        File::delete($tmp);
    }
 
    public function testTmp()
    {
        $data = 'Touch base wiggle room, but mobile friendly, nor blue sky.';
        $filename = File::tmp($data);
        $this->assertTrue(strpos($filename, sys_get_temp_dir()) !== false);
        $this->assertEquals($data, file_get_contents($filename));
    }

    public function testInfo()
    {
        $data = base64_encode(openssl_random_pseudo_bytes(1000000));
        $filename = uniqid() . '.txt';
        $tmpfile = sys_get_temp_dir() . '/' . $filename;
        File::write($tmpfile, $data);
        $expected = [
            'name' => $filename,
            'directory' => sys_get_temp_dir(),
            'timestamp' => fileatime($tmpfile),
            'size' => '1333336',
            'extension' => 'txt',
            'type' => 'text/plain'
        ];
        $this->assertEquals($expected, File::info($tmpfile));
    }
 
    /**
     * @depends testTmp
     */
    public function testRename()
    {
        $data = 'What are the expectations pulling teeth gain traction put a record on and see who dances onward and upward, productize the deliverables and focus on the bottom line.';
        $filename = File::tmp($data);
        $this->assertTrue(File::rename($filename, 'rename.txt'));
        $this->assertEquals($data, file_get_contents(sys_get_temp_dir() . '/rename.txt'));
        unlink(sys_get_temp_dir() . '/rename.txt');
    }

    public function testRenameException()
    {
        $this->expectException(NotFoundException::class);
        File::rename('/foo/bar.txt', 'fozzy.txt');
    }

    /**
     * @depends testTmp
     */
    public function testCopy()
    {
        $data = 'Moving the goalposts thought shower deploy, and that jerk from finance really threw me under the bus, nor critical mass fire up your browser.';
        $filename = File::tmp($data);
        $this->assertTrue(File::copy($filename, 'copied.txt'));
        $this->assertEquals($data, file_get_contents(sys_get_temp_dir() . '/copied.txt'));
        unlink(sys_get_temp_dir() . '/copied.txt');
    }

    public function testCopyException()
    {
        $this->expectException(NotFoundException::class);
        File::copy('/foo/bar.txt', 'fozzy.txt');
    }

    /**
     * @depends testTmp
     */
    public function testMove()
    {
        $data = 'Cannibalize bleeding edge, for net net.';
        $filename = File::tmp($data);
        $this->assertTrue(File::move($filename, sys_get_temp_dir() . '/moved.txt'));
        $this->assertFalse(file_exists($filename));
        $this->assertEquals($data, file_get_contents(sys_get_temp_dir() . '/moved.txt'));
        unlink(sys_get_temp_dir() . '/moved.txt');
    }

    public function testMoveException()
    {
        $this->expectException(NotFoundException::class);
        File::move('/foo/bar.txt', 'fozzy.txt');
    }
    
    public function testPerms()
    {
        $data = 'Not really important';
        $filename = File::tmp($data);
        $this->assertTrue(chmod($filename, 0644));
        $this->assertEquals('0644', File::perms($filename));
    }

    public function testPermsException()
    {
        $this->expectException(NotFoundException::class);
        File::perms('/foo/bar.txt');
    }

    /**
     * Trying to make this test work on multiple systems where user is not known
     */
    public function testOwner()
    {
        $filename = File::tmp('data');
        $owner = File::owner($filename);
        $this->assertMatchesRegularExpression('/^[a-z0-9]+$/i', $owner);
        $this->assertTrue(File::chown($filename, $owner));
    }

    public function testOwnerException()
    {
        $this->expectException(NotFoundException::class);
        File::owner('/foo/bar.txt');
    }

    public function testChmod()
    {
        $tmp = File::tmp('data');
          
        $this->assertTrue(File::chmod($tmp, 0777));
        clearstatcache();
        $this->assertEquals('0777', File::perms($tmp));
    }
  
    public function testGroup()
    {
        $filename = File::tmp('data');
        $group = File::group($filename);
        $this->assertMatchesRegularExpression('/^[a-z0-9]+$/i', $group);
        $this->assertTrue(File::chgrp($filename, $group));
    }
    public function testGroupException()
    {
        $this->expectException(NotFoundException::class);
        File::group('/foo/bar.txt');
    }

    public function testChmodException()
    {
        $this->expectException(NotFoundException::class);
        File::chmod('/foo/file.txt', 0775);
    }

    public function testChownException()
    {
        $this->expectException(NotFoundException::class);
        File::chown('/foo/file.txt', 'some-user');
    }

    public function testChgrpException()
    {
        $this->expectException(NotFoundException::class);
        File::chgrp('/foo/file.txt', 'some-group');
    }
}


tests/FileObjectTest.php
<?php

/**
 * OriginPHP Framework
 * Copyright 2018 - 2021 Jamiel Sharief.
 *
 * Licensed under The MIT License
 * The above copyright notice and this permission notice shall be included in all copies or substantial
 * portions of the Software.
 *
 * @copyright   Copyright (c) Jamiel Sharief
 * @link        https://www.originphp.com
 * @license     https://opensource.org/licenses/mit-license.php MIT License
 */

namespace Origin\Test\FileSystem;

use Origin\Filesystem\FileObject;

class FileObjectTest extends \PHPUnit\Framework\TestCase
{
    public function testAccess()
    {
        $data = ['name' => 'foo.txt','path' => 'folder/subfolder/foo.txt','directory' => 'folder/subfolder', 'size' => 32000, 'timestamp' => strtotime('2019-10-31 14:40')];
        $object = new FileObject($data);

        $this->assertEquals('foo.txt', $object['name']);
        $this->assertEquals('folder/subfolder', $object['directory']);

        $this->assertEquals('foo.txt', $object->name);
        $this->assertEquals('folder/subfolder', $object->directory);

        $this->assertTrue(isset($object['name']));
        $this->assertTrue(isset($object->name));

        $this->assertNull($object['abc']);
        $this->assertNull($object->abc);

        unset($object['name']);
        $object->name = 'bar.txt';
        $this->assertEquals('bar.txt', $object->name);

        unset($object->name);
        $object['name'] = 'bar.txt';
        $this->assertEquals('bar.txt', $object->name);

        // test offsetset
        $object[] = 'foo';
        $this->assertEquals('foo', $object[0]);

        // Need to call this to ensure no errors but cant test it.
        unset($object->furion);

        $this->assertEquals('folder/subfolder/foo.txt', (string)  $object);
    }
}


src/FileObject.php
<?php
/**
 * OriginPHP Framework
 * Copyright 2018 - 2021 Jamiel Sharief.
 *
 * Licensed under The MIT License
 * The above copyright notice and this permission notice shall be included in all copies or substantial
 * portions of the Software.
 *
 * @copyright     Copyright (c) Jamiel Sharief
 * @link         https://www.originphp.com
 * @license       https://opensource.org/licenses/mit-license.php MIT License
 */
declare(strict_types = 1);
namespace Origin\Filesystem;

use ArrayAccess;

/**
 * FileObject for holding file result can work as array or object
 */
class FileObject implements ArrayAccess
{
    private $data = [];

    public function __construct(array $data)
    {
        $this->data = $data;
    }

    public function offsetSet($offset, $value)
    {
        if (is_null($offset)) {
            $this->data[] = $value;
        } else {
            $this->data[$offset] = $value;
        }
    }

    public function offsetExists($offset)
    {
        return isset($this->data[$offset]);
    }

    public function offsetUnset($offset)
    {
        unset($this->data[$offset]);
    }

    public function offsetGet($offset)
    {
        return $this->data[$offset] ?? null;
    }

    public function &__get($key)
    {
        $value = null;
        if (isset($this->data[$key])) {
            $value = &$this->data[$key];
        }

        return $value;
    }

    public function __set($key, $value)
    {
        $this->data[$key] = $value;
    }

    public function __isset($key)
    {
        return isset($this->data[$key]);
    }

    public function __unset($key)
    {
        unset($this->data[$key]);
    }

    public function __debugInfo()
    {
        return $this->data;
    }

    public function __toString()
    {
        return $this->data['path'];
    }
}


src/Folder.php
<?php
/**
 * OriginPHP Framework
 * Copyright 2018 - 2021 Jamiel Sharief.
 *
 * Licensed under The MIT License
 * The above copyright notice and this permission notice shall be included in all copies or substantial
 * portions of the Software.
 *
 * @copyright   Copyright (c) Jamiel Sharief
 * @link        https://www.originphp.com
 * @license     https://opensource.org/licenses/mit-license.php MIT License
 */
declare(strict_types = 1);
namespace Origin\Filesystem;

use Origin\Filesystem\Exception\NotFoundException;
use function Origin\Defer\defer;

class Folder
{
    const MODE = 0775;
    
    /**
     * Creates a folder or folders recursively
     *
     * @param string $directory
     * @param array $options
     *  - recursive: default false. like -p creates all directories in one go
     *  - mode: default 0775. Folder permissions
     * @return boolean
     */
    public static function create(string $directory, array $options = []): bool
    {
        $options += ['recursive' => false,'mode' => self::MODE];
        defer($a, 'umask', umask(0));

        return @mkdir($directory, $options['mode'], $options['recursive']); # use@ No such file or directory
    }

    /**
     * Checks if a directory exists
     *
     * @param string $directory
     * @return boolean
     */
    public static function exists(string $directory): bool
    {
        return is_dir($directory);
    }

    /**
     * Includes a list of files
     *
     * @param string $directory
     * @param array $options Options keys are
     *  - directories: default false, includes directories
     *  - recursive: default false. Recursively gets contents
     * @return array
     */
    public static function list(string $directory, array $options = []): array
    {
        $options += ['directories' => false,'recursive' => false];
        if (self::exists($directory)) {
            $results = [];
            $files = array_diff(scandir($directory), ['.', '..']);
            foreach ($files as $file) {
                if ($options['recursive'] && ! is_file($directory . DIRECTORY_SEPARATOR . $file)) {
                    $results = array_merge($results, static::list($directory . DIRECTORY_SEPARATOR . $file, $options));
                }
                if (! $options['directories'] && ! is_file($directory . DIRECTORY_SEPARATOR . $file)) {
                    continue;
                }
                $stats = stat($directory . DIRECTORY_SEPARATOR . $file);
                $results[] = new FileObject([
                    'name' => $file,
                    'directory' => $directory,
                    'path' => $directory . DIRECTORY_SEPARATOR . $file,
                    'extension' => pathinfo($file, PATHINFO_EXTENSION),
                    'timestamp' => $stats['mtime'],
                    'size' => $stats['size'],
                    'type' => is_dir($directory. DIRECTORY_SEPARATOR . $file) ? 'directory' : 'file',
                ]);
            }

            return $results;
        }
       
        throw new NotFoundException(sprintf('%s does not exist', $directory));
    }

    /**
     * Deletes a directory. For saftey reasons recrusive is disabled by default,
     * since this will delete files etc.
     *
     * @param string $directory
     * @param array $options These are the options can you use:
     *    - recursive: If set to true, it will delete all contents and sub folders
     * @return bool
     */
    public static function delete(string $directory, array $options = []): bool
    {
        $options += ['recursive' => false];
        if (self::exists($directory)) {
            if ($options['recursive']) {
                $files = array_diff(scandir($directory), ['.', '..']);
                foreach ($files as $filename) {
                    if (is_dir($directory . DIRECTORY_SEPARATOR . $filename)) {
                        self::delete($directory . DIRECTORY_SEPARATOR . $filename, $options);
                        continue;
                    }
                    @unlink($directory . DIRECTORY_SEPARATOR . $filename);
                }
            }

            return @rmdir($directory);
        }
        
        throw new NotFoundException(sprintf('%s does not exist', $directory));
    }

    /**
     * Renames a directory
     *
     * @param string $directory full patth e.g. /var/www/tmp/my_project
     * @param string $to  directory name. project_name
     * @return boolean
     */
    public static function rename(string $directory, string $to): bool
    {
        if (self::exists($directory)) {
            if (strpos($to, DIRECTORY_SEPARATOR) === false) {
                $to = pathinfo($directory, PATHINFO_DIRNAME) . DIRECTORY_SEPARATOR . $to;
            }

            return @rename($directory, $to);
        }
        throw new NotFoundException(sprintf('%s could not be found', $directory));
    }

    /**
     * Moves a directory
     *
     * @param string $source /var/www/tmp/docs
     * @param string $destination /var/www/tmp/documents
     * @return bool
     */
    public static function move(string $source, string $destination): bool
    {
        if (self::exists($source)) {
            if (strpos($destination, DIRECTORY_SEPARATOR) === false) {
                $destination = pathinfo($source, PATHINFO_DIRNAME) . DIRECTORY_SEPARATOR . $destination;
            }

            return @rename($source, $destination);
        }
        throw new NotFoundException(sprintf('%s could not be found', $source));
    }

    /**
     * Copies a directory
     *
     * @param string $source /var/www/tmp/my_project
     * @param string $destination project_name or /var/www/tmp/project_name
     * @param array $options The options array supports the following keys
     *   - recursive: default true. recursively copy the contents of folder
     * @return boolean
     */
    public static function copy(string $source, string $destination, array $options = []): bool
    {
        $options += ['recursive' => true];

        if (self::exists($source)) {
            if (strpos($destination, DIRECTORY_SEPARATOR) === false) {
                $destination = pathinfo($source, PATHINFO_DIRNAME) . DIRECTORY_SEPARATOR . $destination;
            }

            if (! is_dir($destination)) {
                @mkdir($destination, self::MODE);
            }
           
            $files = array_diff(scandir($source), ['.', '..']);
            foreach ($files as $filename) {
                if ($options['recursive'] && is_dir($source . DIRECTORY_SEPARATOR . $filename)) {
                    self::copy($source . DIRECTORY_SEPARATOR . $filename, $destination . DIRECTORY_SEPARATOR . $filename, $options);
                    continue;
                }
                @copy($source . DIRECTORY_SEPARATOR . $filename, $destination . DIRECTORY_SEPARATOR . $filename);
            }

            return self::exists($destination);
        }
        throw new NotFoundException(sprintf('%s could not be found', $source));
    }

    /**
    * Gets or sets the permissions (mode) for a directory
    *
    * @param string $directory filename with full path
    * @return string
    */
    public static function mode(string $directory): string
    {
        if (self::exists($directory)) {
            return (string) substr(sprintf('%o', fileperms($directory)), -4);
        }
        throw new NotFoundException(sprintf('%s could not be found', $directory));
    }

    /**
    * Alias for mode. Gets the mode for a file aka permissions
    *
    * @param string $directory
    * @return string|null
    */
    public static function perms(string $directory): ?string
    {
        return static::mode($directory);
    }

    /**
      * Gets the owner of directory
      *
      * @param string $directory filename with full path
      * @return string|null
      */
    public static function owner(string $directory): ?string
    {
        if (self::exists($directory)) {
            return posix_getpwuid(fileowner($directory))['name'];
        }
        throw new NotFoundException(sprintf('%s could not be found', $directory));
    }

    /**
     * Gets the group that the directory belongs to.
     *
     * @param string $directory filename with full path
     * @return string|null
     */
    public static function group(string $directory): ?string
    {
        if (self::exists($directory)) {
            return posix_getgrgid(filegroup($directory))['name'];
        }
        throw new NotFoundException(sprintf('%s could not be found', $directory));
    }

    /**
    * Changes the permissions of directory
    *
    * @param string $directory filename with full path
    * @param int $mode e.g 0755 (remember 0 infront)
     * @param array $options (recursive default false)
     *  - recursive: If set to true, it will delete all contents and sub folders
    * @return bool
    */
    public static function chmod(string $directory, int $mode, array $options = []): bool
    {
        $options += ['recursive' => false];

        if (self::exists($directory)) {
            if ($options['recursive']) {
                $files = array_diff(scandir($directory), ['.', '..']);
                foreach ($files as $filename) {
                    if (is_dir($directory . DIRECTORY_SEPARATOR . $filename)) {
                        self::chmod($directory . DIRECTORY_SEPARATOR . $filename, $mode, $options);
                        continue;
                    }
                    @chmod($directory . DIRECTORY_SEPARATOR . $filename, $mode);
                }
            }

            return @chmod($directory, $mode);
        }
        throw new NotFoundException(sprintf('%s could not be found', $directory));
    }

    /**
     * Changes the owner of the directory
     *
     * @param string $directory filename with full path
     * @param string $user  e.g. root, www-data
     * @param array $options Support options keys are
     *  - recursive: default false. If set to true, it will delete all contents and sub folders
     * @return bool
     */
    public static function chown(string $directory, string $user, array $options = []): bool
    {
        $options += ['recursive' => false];
        if (self::exists($directory)) {
            if ($options['recursive']) {
                $files = array_diff(scandir($directory), ['.', '..']);
                foreach ($files as $filename) {
                    if (is_dir($directory . DIRECTORY_SEPARATOR . $filename)) {
                        self::chown($directory . DIRECTORY_SEPARATOR . $filename, $user, $options);
                        continue;
                    }
                    @chown($directory . DIRECTORY_SEPARATOR . $filename, $user);
                }
            }

            return @chown($directory, $user);
        }
        throw new NotFoundException(sprintf('%s could not be found', $directory));
    }

    /**
    * Changes the group that the directory belongs to
    *
    * @param string $directory filename with full path
    * @param string $group  e.g. root, www-data
    * @param array $options (recursive default false)
    *  - recursive: If set to true, it will delete all contents and sub folders
    * @return bool
    */
    public static function chgrp(string $directory, string $group = null, array $options = []): bool
    {
        $options += ['recursive' => false];
        if (self::exists($directory)) {
            if ($options['recursive']) {
                $files = array_diff(scandir($directory), ['.', '..']);
                foreach ($files as $filename) {
                    if (is_dir($directory . DIRECTORY_SEPARATOR . $filename)) {
                        self::chgrp($directory . DIRECTORY_SEPARATOR . $filename, $group, $options);
                        continue;
                    }
                    @chgrp($directory . DIRECTORY_SEPARATOR . $filename, $group);
                }
            }

            return @chgrp($directory, $group);
        }
        throw new NotFoundException(sprintf('%s could not be found', $directory));
    }
}
